boa/src/builtins/object/mod.rs

```diff
@@ -45,8 +48,214 @@ mod tests;
 /// Static `prototype`, usually set on constructors as a key to point to their respective prototype object.
 pub static PROTOTYPE: &str = "prototype";
 
+/// This trait allows Rust types to be passed around as objects.
+///
+/// This is automatically implemented, when a type implements `Debug`, `Any` and `Trace`.
+pub trait NativeObject: Debug + Any + Trace {
+    fn as_any(&self) -> &dyn Any;
+    fn as_mut_any(&mut self) -> &mut dyn Any;
+}
+
+impl<T: Any + Debug + Trace> NativeObject for T {
+    fn as_any(&self) -> &dyn Any {
+        self as &dyn Any
+    }
+
+    fn as_mut_any(&mut self) -> &mut dyn Any {
+        self as &mut dyn Any
+    }
+}
+
+/// Native class.
+pub trait Class: NativeObject + Sized {
+    /// The binding name of the object.
+    const NAME: &'static str;
+    /// The amount of arguments the class `constructor` takes, default is `0`.
+    const LENGTH: usize = 0;
+    /// The attibutes the class will be binded with, default is `writable`, `enumerable`, `configurable`.
+    const ATTRIBUTE: Attribute = Attribute::all();
+
+    /// The constructor of the class.
+    fn constructor(this: &Value, args: &[Value], ctx: &mut Interpreter) -> Result<Self>;
+
+    /// Initializes the internals and the methods of the class.
+    fn methods(class: &mut ClassBuilder<'_>) -> Result<()>;
```

Comment:
Should this method be called `init` instead of `methods` since its not just the methods that are being initialized.

Reply:
It makes sense, yes!